src/Components/WebAssembly/WebAssembly/src/Rendering/WebAssemblyDispatcher.cs

```diff
@@ -0,0 +1,163 @@
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Rendering;
+
+// When Blazor is deployed with multi-threaded runtime, WebAssemblyDispatcher will help to dispatch all Blazor JS interop calls to the main thread.
+// This is necessary because all JS objects have thread affinity. They are only available on the thread (WebWorker) which created them.
+// Also DOM is only available on the main (browser) thread.
+// Because all of the Dispatcher.InvokeAsync methods return Task, we don't need to propagate errors via OnUnhandledException handler
+internal sealed class WebAssemblyDispatcher : Dispatcher
+{
+    public static readonly Dispatcher Instance = new WebAssemblyDispatcher();
+    private readonly SynchronizationContext _mainSyncContext;
+
+    private WebAssemblyDispatcher()
+    {
+        Debug.Assert(Environment.CurrentManagedThreadId == 1); // WebAssemblyDispatcher should be initialized on the main thread.
+
+        // capture the JSSynchronizationContext from the main thread, which runtime already installed.
+        _mainSyncContext = SynchronizationContext.Current!;
+
+        Debug.Assert(_mainSyncContext != null);
+        Debug.Assert(_mainSyncContext.GetType().FullName == "System.Runtime.InteropServices.JavaScript.JSSynchronizationContext");
+    }
+
+    public override bool CheckAccess() => SynchronizationContext.Current == _mainSyncContext;
+
+    public override Task InvokeAsync(Action workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (CheckAccess())
+        {
+            // this branch executes on correct thread and solved JavaScript objects thread affinity
+            // but it executes out of order, if there are some pending jobs in the _mainSyncContext already, same as RendererSynchronizationContextDispatcher
+            workItem();
+            // it can throw synchronously, same as RendererSynchronizationContextDispatcher
+            return Task.CompletedTask;
+        }
+
+        var tcs = new TaskCompletionSource();
+
+        // RendererSynchronizationContext doesn't need to deal with thread affinity and so it could execute jobs on calling thread as optimization.
+        // we could not do it for WASM/JavaScript, because we need to solve for thread affinity of JavaScript objects, so we always Post into the queue.
+        _mainSyncContext.Post(static (object? o) =>
+        {
+            var state = ((TaskCompletionSource tcs, Action workItem))o!;
+            try
+            {
+                state.workItem();
+                state.tcs.SetResult();
+            }
+            catch (Exception ex)
+            {
+                state.tcs.SetException(ex);
+            }
+        }, (tcs, workItem));
+
+        return tcs.Task;
+    }
+
+    public override Task<TResult> InvokeAsync<TResult>(Func<TResult> workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (CheckAccess())
+        {
+            return Task.FromResult(workItem());
```

Comment:
this means if workItem throws, it will just bubble out instead of wrapping the exception in the Task return value. Is that intended? It feels like it would be better if exceptions always propagated out through the task

Reply:
This is intentional for consistency with https://github.com/dotnet/aspnetcore/blob/0ed0bcdbfd0e0ca1617179e495ef893e36cb332c/src/Components/Components/src/Rendering/RendererSynchronizationContextDispatcher.cs#L46